deployment/terraform/keycloak.go

```diff
@@ -0,0 +1,356 @@
+	}
+
+	// Wait for keycloak to start
+	url := fmt.Sprintf("http://%s:8080/health", t.output.KeycloakServer.PublicDNS)
+	timeout := time.After(120 * time.Second) // yes, is **that** slow
+	for {
+		resp, err := http.Get(url)
+		if err != nil {
+			// Avoid error spamming
+			if !errors.Is(err, syscall.ECONNREFUSED) {
+				mlog.Error("Failed to connect to keycloak", mlog.Err(err))
+				continue
+			}
+		}
+
+		if err == nil {
+			resp.Body.Close()
+			// Keycloak is ready
+			if resp.StatusCode == http.StatusOK {
+				break
+			}
+		}
+
+		mlog.Info("Keycloak service not up yet, waiting...")
+		select {
+		case <-timeout:
+			return errors.New("timeout: keycloak service is not responding")
+		case <-time.After(5 * time.Second):
+		}
+	}
+
+	// Authenticate as admin to execute keycloak commands
+	_, err = sshc.RunCommand(fmt.Sprintf("%s/kcadm.sh config credentials --server http://127.0.0.1:8080 --user %s --password %s --realm master", keycloakBinPath, t.config.ExternalAuthProviderSettings.KeycloakAdminUser, t.config.ExternalAuthProviderSettings.KeycloakAdminPassword))
+	if err != nil {
+		return fmt.Errorf("failed to authenticate keycload admin: %w", err)
+	}
+
+	// Disable SSL requirement on master realm to allow http connections on the web interface
+	_, err = sshc.RunCommand(keycloakBinPath + "/kcadm.sh update realms/master -s sslRequired=NONE")
+	if err != nil {
+		return fmt.Errorf("failed to disable ssl requirement: %w", err)
+	}
+
+	// Populate users
+	if t.config.ExternalAuthProviderSettings.GenerateUsersCount > 0 {
+		if err := t.populateKeycloakUsers(sshc); err != nil {
+			return fmt.Errorf("failed to populate keycloak users: %w", err)
+		}
+
+		mlog.Info("Overriding the users file path with the generated one from keycloak")
+		t.config.UsersFilePath = t.getAsset("keycloak-users.txt")
+	}
+
+	mlog.Info("Keycloak configured")
+
+	return nil
+}
+
+// populateKeycloakUsers creates users in keycloak and writes their credentials to a users file.
+// It will use the `GenerateUsersCount` configuration option to create as many users as the configuration
+// specifies. If the users file already exists and has the expected number of users, it will skip user creation.
+// If the file has less users than expected, it will start creating users from the next number by counting the
+// number of lines in the file.
+// If the file has more users than expected, it will log and skip the user creation.
+func (t *Terraform) populateKeycloakUsers(sshc *ssh.Client) error {
+	keycloakBinPath := "/opt/keycloak/keycloak-" + t.config.ExternalAuthProviderSettings.KeycloakVersion + "/bin"
+	usersTxtPath := t.getAsset("keycloak-users.txt")
+	startNumber := 1
+
+	// Check if users file exists and has the expected number of users. If the file has less users than expected,
+	// we will start creating users from the next number, otherwise we will skip user creation.
+	if _, err := os.Stat(usersTxtPath); err == nil || os.IsExist(err) {
```

Comment:
Why would `Stat` return an error if the file exists?

Reply:
I think this was supposed to be `os.IsNotExist` in a previous apporach or something else with a different logic and I just inherited it here, this doesn't make any sense at the moment.